service/config_watcher.go

```diff
@@ -29,49 +27,42 @@ type configWatcher struct {
 	cfg     *config.Config
 	ret     configmapprovider.Retrieved
 	watcher chan error
-	stopWG  sync.WaitGroup
 }
 
 func newConfigWatcher(ctx context.Context, set CollectorSettings) (*configWatcher, error) {
-	ret, err := set.ConfigMapProvider.Retrieve(ctx)
+	cm := &configWatcher{watcher: make(chan error, 1)}
+
+	ret, err := set.ConfigMapProvider.Retrieve(ctx, cm.onChange)
 	if err != nil {
 		return nil, fmt.Errorf("cannot retrieve the configuration: %w", err)
 	}
 
 	var cfg *config.Config
-	if cfg, err = set.ConfigUnmarshaler.Unmarshal(ret.Get(), set.Factories); err != nil {
+	m, err := ret.Get(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get the configuration: %w", err)
+	}
+	if cfg, err = set.ConfigUnmarshaler.Unmarshal(m, set.Factories); err != nil {
 		return nil, fmt.Errorf("cannot unmarshal the configuration: %w", err)
 	}
 
 	if err = cfg.Validate(); err != nil {
 		return nil, fmt.Errorf("invalid configuration: %w", err)
 	}
 
-	cm := &configWatcher{cfg: cfg, ret: ret, watcher: make(chan error, 1)}
-	// If the retrieved value is watchable start a goroutine watching for updates.
-	if watchable, ok := ret.(configmapprovider.WatchableRetrieved); ok {
-		cm.stopWG.Add(1)
-		go func() {
-			defer cm.stopWG.Done()
-			err = watchable.WatchForUpdate()
-			if errors.Is(err, configsource.ErrSessionClosed) {
-				// This is the case of shutdown of the whole collector server, nothing to do.
-				return
-			}
-			cm.watcher <- err
-		}()
-	}
+	cm.cfg = cfg
+	cm.ret = ret
 
 	return cm, nil
 }
 
-func (cm *configWatcher) close(ctx context.Context) error {
-	defer func() { close(cm.watcher) }()
-	// If the retrieved value is watchable start a goroutine watching for updates.
-	if watchable, ok := cm.ret.(configmapprovider.WatchableRetrieved); ok {
-		return watchable.Close(ctx)
+func (cm *configWatcher) onChange(event *configmapprovider.ChangeEvent) {
+	if event.Error == nil || event.Error != configsource.ErrSessionClosed {
+		cm.watcher <- event.Error
 	}
+}
 
-	cm.stopWG.Wait()
-	return nil
+func (cm *configWatcher) close(ctx context.Context) error {
+	defer func() { close(cm.watcher) }()
```

Comment:
Not sure if usage of `defer` still makes sense here.

Reply:
Right, it doesn't. Removed.